### PR TITLE
extension: allow the dynamic argument length of the extension's custom function

### LINKED ISSUE #38707: allow the dynamic argument length of the extension's custom function

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #38707

Problem Summary:

Add a new attribute `DynamicArgsLen` in `FunctionDef` to support dynamic argument length. The default value is 0, it means the argument length is fixed. When we set it to n, it means the argument len in `[len(args) - n, len(args)]` is allowed.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: extension/function.go [hunk 1]
```diff
@@ -29,9 +30,14 @@ type FunctionContext interface {
...
 type FunctionDef struct {
-	Name   string
+	// Name is the function's name
+	Name string
+	// EvalTp is the type of the return value
 	EvalTp types.EvalType
+	// ArgTps is the argument types
 	ArgTps []types.EvalType
+	// DynamicArgsLen is the length of the dynamic args
+	DynamicArgsLen int
```
Comment (xhebox) [inline extension/function.go:40]: Maybe we can call it optional argument length.

**Review by xhebox — APPROVED** [review]

**Pull request merged**